### DISCUSSION TITLE
feat(ErrorLogger): Enhance the global exception component function

### DISCUSSION THREAD

**Discussion by izanhzh** [dotnetcore/BootstrapBlazor#4036] [Q&A] [answered]
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Is your feature request related to a problem? Please describe the problem.

我在一个项目中参照了`ErrorLogger`的实现，但是最终效果不太符合我预期，我也看到有人遇到了类似的问题https://github.com/dotnetcore/BootstrapBlazor/issues/217
我希望能够增强`ErrorLogger`的全局异常处理能力：
1. 能够自动处理类似`OnAfterRenderAsync`等这种生命周期中触发的异常，这种异常会导致页面重新渲染，重新触发异常，不断循环
2. 能够自动处理弹窗内的错误，弹窗出现错误也会导致不断循环

### Describe the solution you'd like

经过一番折腾，我能够明白这是Blazor自身全局处理不够强导致，`ErrorBoundary` 基本不可能做到这种支持，最终我选择了一个妥协一点的方案：
1. 检测是否触发循环错误渲染，如果是则采用默认的方案，进入错误页面，打断循环，`OnAfterRenderAsync`这些生命周期中，应当尽量避免异常，或者手动处理异常，但是如果我们想要全局日志记录等，还是需要封装组件统一处理的支持
2. 检测错误弹窗是否正确渲染出来

以下是我折腾出的代码：
```csharp
 public class ExceptionHandler : ErrorBoundaryBase
 {
     [Inject]
     private IErrorBoundaryLogger? ErrorBoundaryLogger { get; set; }

     private bool firstRender = true;
     private bool showExceptionMessagePopover = false;
     private int errorRenderCount = 0;

     protected override async Task OnErrorAsync(Exception exception)
     {
         showExceptionMessagePopover = true;
         errorRenderCount++;
         await ErrorBoundaryLogger!.LogErrorAsync(exception);
     }

     protected override void OnAfterRender(bool firstRender)
     {
         this.firstRender = firstRender;
     }

     protected override void BuildRenderTree(RenderTreeBuilder builder)
     {
         int sequence = 0;
         if (CurrentException is null)
         {
             builder.AddContent(sequence++, ChildContent);
         }
         else if (ErrorContent is not null)
         {
             builder.AddContent(sequence++, ErrorContent(CurrentException));
         }
         else
         {
             if (firstRender || errorRenderCount> 2)
             {
                 builder.OpenElement(sequence++, "div");
                 builder.AddAttribute(sequence++, "class", "blazor-error-boundary");
                 builder.CloseElement();
             }
             else
             {
                 builder.AddContent(sequence++, ChildContent);
                 builder.OpenComponent<ExceptionMessagePopover>(3);
                 builder.AddComponentParameter(sequence++, "Open", showExceptionMessagePopover);
                 builder.AddComponentParameter(sequence++, "OpenChanged", EventCallback.Factory.Create<bool>(this, SetShowExceptionMessagePopover));
                 builder.AddComponentParameter(sequence++, "Exception", CurrentException);
                 builder.CloseComponent();
             }
         }
     }
     private void SetShowExceptionMessagePopover(bool value)
     {
         if (showExceptionMessagePopover && !value)//满足此条件，说明错误提示框成功渲染出来了，没有再触发其他异常，这里重置掉相关状态
         {
             errorRenderCount= 0;
             Recover();
         }
         showExceptionMessagePopover = value;
     }
 }
```

**Comment by bb-auto** [comment]
@izanhzh Thank you for contacting us. We will give feedback later.

**Comment by ArgoZhang** [comment]
@izanhzh 请提供一下示例代码啊。上传个工程给我。大佬的意思是让我们的组件用你这套逻辑对吗？

**Comment by izanhzh** [comment]
没有没有，你才是真大佬，我在上面提供了我折腾的一个代码，是想看看有没有参考价值，或者你们有没有更好的办法，目前blazor的这个全局处理老是差那么点意思，如果大佬觉得这个建议不合适，可以关掉这个feat

**Comment by ArgoZhang** [comment]
@izanhzh 别紧张，能提高组件性能或者功能，我们都乐意合并的。需要示例比如哪里不满意？你这个代码增强了哪里？

**Comment by izanhzh** [comment]
我希望能够自动跳出循环错误的情况，让它能够应对全部地方抛出的异常，如果是像初始化时触发的异常，实在没办法弹窗，就退而求其次，进入错误页面，包括弹窗提示的时候有异常，也进入错误页面

**Reply by ArgoZhang** [reply]
其实我还是没懂你的需求。你什么都不做，就应该是进去 Error 页面啊。这里 net8 跟非 net8 处理有一些不一样而已

**Reply by izanhzh** [reply]
哈哈，我知道为啥你没太懂我的需求了，我早上试了下你这个组件，我之前只是参照你的代码来实现我的需求，还没真正用过你这个组件，我之前理解错了你这个组件的作用。

我在系统开发中，喜欢自定义BusinessException/UserFriendlyException，这类异常是我业务逻辑校验时故意抛出来的，我期望UI能够自动识别这种异常进行友好的弹窗提示用户，这种异常不属于未知异常范畴。因此，我需要Blazor有一个统一的异常处理组件，当检测到是BusinessException/UserFriendlyException时，UI弹窗提示用户即可，不需要跳转到Error页面去。当出现未知异常时，UI弹窗也要进行用户友好化提示，比如统一提示：系统发生了一个未知错误，请联系系统管理人员，错误码：xxx。

这样做有两个目的：
1. 简化业务逻辑错误时提示的代码，写业务逻辑代码的按约定抛BusinessException/UserFriendlyException，不用关心UI怎么去提示用户（甚至是提示时的多语言处理等），不需要重复写弹窗提示代码。
2. 出现未知异常不会导致程序崩溃，隐藏异常细节，用户看到比较友好的提示，用户截图告知我们错误UI，我们通过错误码就能知道细节

我早上试了下你这个组件，拿微软默认那个模板的Conuter页面来讲
```csharp
@page "/counter"

<PageTitle>Counter</PageTitle>

<ErrorLogger>
    <h1>Counter</h1>

    <p role="status">Current count: @currentCount</p>

    <button class="btn btn-primary" @onclick="IncrementCount">Click me</button>
</ErrorLogger>

@code {
    private int currentCount = 0;

    private void IncrementCount()
    {
        currentCount++;

        var a = 0;
        _ = 1 / a;
    }
}
```
当我点击时，会触发非BusinessException/UserFriendlyException异常，我先定义为是未知的异常，由开发人员不小心产生的，我想要弹窗提示，用了你这个组件，它能弹窗提示，但是它会进入错误页面，当我发布部署后，虽然不会进入错误页面，但是这个页面也不能再继续操作了：
![image](https://github.com/user-attachments/assets/8936aa4d-5fde-496d-87a6-dd30f913753e)
![image](https://github.com/user-attachments/assets/ba6c9cf3-7456-423b-9f06-b7a758390959)

**Reply by izanhzh** [reply]
但是我看你的文档那边提供的效果又是符合我预期的，很奇怪，我自己试一旦触发异常，这个页面就不能再继续用了，只能Reload。

我之前是在.net8 auto呈现模式下参照你这个组件代码去实现我要的效果，我发现如果异常不是在用户点击按钮等操作产生的，而是在OnInitialized这种地方不小心产生，的确是无法做到弹窗而不进入错误页面，这种情况我能够妥协接受让它进入错误页面，错误页面进行友好提示

**Reply by ArgoZhang** [reply]
这个组件的最大功能就是防止页面不能用，是出于这个功能点设计的。报错后底下出现黄条是不能容忍的。我猜肯定是哪里没弄对，但是我又猜不出来

**Reply by izanhzh** [reply]
有QQ/微信群么，你文档底部的社区 联系我们 这些按钮点击没有效果

**Reply by izanhzh** [reply]
[BlazorApp2.zip](https://github.com/user-attachments/files/16593988/BlazorApp2.zip)

**Reply by ArgoZhang** [reply]
QQ 群 795206915

**Reply by izanhzh** [reply]
满员了，我代码压缩发出来了

**Reply by ArgoZhang** [reply, marked as answer]
代码我看了。问题如下 

```
<ErrorLogger>
    <h1>Counter</h1>

    <p role="status">Current count: @currentCount</p>

    <button class="btn btn-primary" @onclick="IncrementCount">Click me</button>
</ErrorLogger>
```

不能用原生的 button 这样拦不住
